nexus/src/app/sagas/region_replacement_start.rs

```diff
@@ -0,0 +1,1259 @@
+    let datasets_and_regions = osagactx
+        .datastore()
+        .get_allocated_regions(db_region.volume_id())
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    Ok(datasets_and_regions)
+}
+
+async fn srrs_alloc_new_region(
+    sagactx: NexusActionContext,
+) -> Result<Vec<(db::model::Dataset, db::model::Region)>, ActionError> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+
+    let opctx = crate::context::op_context_for_saga_action(
+        &sagactx,
+        &params.serialized_authn,
+    );
+
+    // Look up the existing region: we want to duplicate it but at another
+    // physical disk
+    let db_region = osagactx
+        .datastore()
+        .get_region(params.request.old_region_id)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    // Request an additional region for this volume: THRESHOLD + 1 is required
+    // in order to have the proper redundancy. It's important _not_ to delete
+    // the existing region first, as (if it's still there) then the Crucible
+    // agent could reuse the allocated port and cause trouble.
+    let datasets_and_regions = osagactx
+        .datastore()
+        .arbitrary_region_allocate_direct(
+            &opctx,
+            db_region.volume_id(),
+            db_region.block_size().to_bytes(),
+            db_region.blocks_per_extent(),
+            db_region.extent_count(),
+            &params.allocation_strategy,
+            // Note: this assumes that previous redundancy is
+            // REGION_REDUNDANCY_THRESHOLD, and that region replacement will
+            // only be run for volumes that start at this redundancy level.
+            REGION_REDUNDANCY_THRESHOLD + 1,
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    Ok(datasets_and_regions)
+}
+
+fn find_only_new_region(
+    log: &Logger,
+    existing_datasets_and_regions: Vec<(db::model::Dataset, db::model::Region)>,
+    new_datasets_and_regions: Vec<(db::model::Dataset, db::model::Region)>,
+) -> Option<(db::model::Dataset, db::model::Region)> {
+    // Only filter on whether or not a Region is in the existing list! Datasets
+    // can change values (like size_used) if this saga interleaves with other
+    // saga runs of the same type.
+    let mut dataset_and_region: Vec<(db::model::Dataset, db::model::Region)> =
+        new_datasets_and_regions
+            .into_iter()
+            .filter(|(_, r)| {
+                !existing_datasets_and_regions.iter().any(|(_, er)| er == r)
+            })
+            .collect();
+
+    if dataset_and_region.len() != 1 {
+        error!(
+            log,
+            "find_only_new_region saw dataset_and_region len {}: {:?}",
+            dataset_and_region.len(),
+            dataset_and_region,
+        );
+
+        None
+    } else {
+        dataset_and_region.pop()
+    }
+}
+
+async fn srrs_alloc_new_region_undo(
+    sagactx: NexusActionContext,
```

Comment:
You're correct - the region creation code will wait for a known result (created or otherwise) before proceeding, and so will the region deletion, and therefore the saga will only proceed when a result of a request is known.